files/private-chef-ctl-commands/key-control.rb

```diff
@@ -14,345 +14,418 @@
-    # there are reasons?
-    db_user = running_config['private_chef']['postgresql']['sql_user']
-    db_password = running_config['private_chef']['postgresql']['sql_password']
-    db_host = running_config['private_chef']['postgresql']['listen_address']
-
-    # Set convert_infinite_timestamps to a float so that Sequel
-    # doesn't crash out on querying an infinity timestamp.
-    # See this commit for more detail:
-    # https://github.com/jeremyevans/sequel/commit/2440e98b4a7063a90d8337e57b1c3e3d3a1cbd0b
-    @db = Sequel.connect(:adapter => 'postgres', :host => db_host,
-                         :database => 'opscode_chef', :user => db_user,
-                         :password => db_password, :convert_infinite_timestamps => :float)
-
-  end
-end
-
-class PublicKeyOutputter
-  def output_key_results(results, hide_public_keys, database)
-    results.each do |row|
-      # get expiration date from keys table as
-      # results variable is result of keys_by_name view
-      expires_at = database.db[:keys].where(:id => row[:id]).where(:key_name => row[:key_name]).first[:expires_at]
-      puts "\nkey_name: #{row[:key_name]}"
-      puts "expires_at: #{expires_at}"
-      unless hide_public_keys
-        puts "public_key:"
-        puts "#{row[:public_key]}"
+Chef::Config.from_file("/etc/opscode/pivotal.rb")
+@chef_rest = Chef::REST.new(Chef::Config[:chef_server_root])
+
+class Helper
+  # since optparse is stupid, make sure that the "argument" passed into the currently
+  # parsed argflag isn't the next argflag. if it is an argflag, they passed nothing
+  # to the current argument, so throw  the proper error.
+  def catch_no_arg(arg_list, arg, arg_input)
+    arg_list.each do |arg_element|
+      if arg_input.strip == arg_element
+        STDERR.puts "Error: Missing valid input for argument #{arg}."
+        exit 1
       end
     end
   end
-end
 
-class ArgParser
-  def find_arg_after_identifier(args, identifier)
-    result = nil
-    args.each_with_index do |arg, i|
-      if arg == identifier
-        result = args[i+1]
-        break
-      end
+  def read_and_check_key(key_path, error_msg)
+    begin
+      key = File.read(key_path)
+    rescue
+      STDERR.puts error_msg
+      exit 1
     end
-    result
-  end
-end
 
-class AddKey
-  def check_org_args(cmd_args, usage_msg)
-    unless cmd_args.length == 2 || \
-      (cmd_args.length == 4 && (cmd_args[2] == "--expiration-date" || cmd_args[2] == "--key-name")) || \
-      (cmd_args.length == 6 && ((cmd_args[2] == "--expiration-date" && cmd_args[4] == "--key-name") || (cmd_args[4] == "--expiration-date" && cmd_args[2] == "--key-name")))
-      STDERR.puts usage_msg
-      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
-      STDERR.puts "Usage: Default name used is the fingerprint of the key passed"
+    unless /-----BEGIN PUBLIC KEY-----/.match(key)
```

Comment:
Probably worth accepting. I think the specific version of RSA we are using expects that header, but that regex is probably too strict.